proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,121 @@
+Therefore, most control planes are forced to simply act as if the resource does not exist. In this situation, the client does not know why the control plane was unable to send the resource, so it cannot report useful information to human operators (e.g., via logs or request failure messages), which makes debugging challenging.
+
+Note that while the incremental protocol variants can explicitly indicate to the client that a subscribed resource does not exist, the SotW protocol variants require the client to use a 15-second does-not-exist timeout. This adds an unnecessary delay before the client can react to the error. It has also led to cases where control plane slowness has caused clients to incorrectly react as if a resource does not exist.
+
+### Related Proposals:
+
+* [TP1-xds-transport-next.md](TP1-xds-transport-next.md)
+
+## Proposal
+
+The proposal introduces two new fields to the DiscoveryResponse proto. The first, `resource_errors`, will detail Resource-specific errors and be included in both SotW and Incremental xDS protocols. The second, `removed_resources`, will specifically identify resources that no longer exist, addressing a current gap in the SotW protocol's signaling capabilities. 
+
+```textproto
+// New Proto Message
+// Contains detailed error detail for a single resource
+message ResourceError {
+ ResourceName resource_name = 1;
+
+ google.rpc.Status error_detail = 2;
+}
+
+message DiscoveryResponse {
+
+// The version of the response data.
+string version_info = 1;
+
+….
+…
+
+
+ // The control plane instance that sent the response.
+  config.core.v3.ControlPlane control_plane = 6;
+
+// NEW_FIELD
+// An optional repeated list of error details that the control plane 
+// can provide to notify the client of issues for the resources that 
+// are subscribed.
+// This allows the xDS management server to provide optional 
+// notifications in case of unavailability, permissions errors 
+// without the client having to wait for the `config fetch` timeout.
+repeated ResourceError resource_errors = N;
+...
+}
+
+message DeltaDiscoveryResponse {
+  …
+
+  // NEW FIELD
+  repeated ResourceError resource_errors = N;
+}
+```
+
+### Protocol Behavior
+Clients that support this mechanism will use this additional field to obtain error information for resources that the xDS server couldn’t provide. When this message is received, the xDS client should cancel any resource timers for the indicated resource. The client should then handle the error based on the status code, as follows (using [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) terminology):
+
+  * The following codes will be interpretted by clients as transient failures, meaning that the client MUST continue to use previously cached versions of the resource if it has them: UNAVAILABLE, INTERNAL, UNKNOWN.
```

Comment:
I can think of several reasons why the client might later see an error after having previously seen a valid resource.  For example, maybe the ADS stream gets closed and the client winds up reconnecting to a different xDS server instance that has a different view of the resources.

Reply:
BTW, I should clarify that this doesn't affect the cache at the xDS transport protocol layer; it's really referring to state kept at the data model layer.  The watcher still needs to cache the previous resource in either case, but an individual watcher (e.g., in a Cluster implementation) may decide to that it doesn't want to continue to use that cached data if it sees a particular error.

Reply:
This made me realize that we also need to define the CSDS behavior here.  That doesn't affect Envoy, since Envoy doesn't support CSDS yet, but gRPC does.

I think we need to add a new value to the [`ClientConfigStatus` enum](https://github.com/envoyproxy/envoy/blob/22e7328305b8fc682644e1bfbbc33f93b856993d/api/envoy/service/status/v3/csds.proto#L60), let's call it `CLIENT_RECEIVED_ERROR`.  This should be treated basically the same as `CLIENT_NACKED`: it will provide the previously cached version of the resource but also indicate the server error that we got in the [`error_state` field](https://github.com/envoyproxy/envoy/blob/22e7328305b8fc682644e1bfbbc33f93b856993d/api/envoy/service/status/v3/csds.proto#L170).

Reply:
> I can think of several reasons why the client might later see an error after having previously seen a valid resource. For example, maybe the ADS stream gets closed and the client winds up reconnecting to a different xDS server instance that has a different view of the resources.

The issue is that the error may be intermittent, because the newly connected server isn't aware of the resource.



> BTW, I should clarify that this doesn't affect the cache at the xDS transport protocol layer; it's really referring to state kept at the data model layer. The ~watcher~ transport protocol layer still needs to cache the previous resource in either case, but an individual watcher (e.g., in a Cluster implementation) may decide to that it doesn't want to continue to use that cached data if it sees a particular error.

IMHO if the decision of whether a resource should or should not be used is left to the resource owner that would be a violation of the protocol and may lead to inconsistencies between different data-plane implementations.

Reply:
@anicr7, I think we still need to include the CSDS change that I suggested in a comment earlier in this thread:

https://github.com/cncf/xds/pull/107#discussion_r1853159726

Reply:
Done